src/NuGet.Services.EndToEnd/Support/Clients/V2V3SearchClient.cs

```diff
@@ -323,37 +326,93 @@ private async Task<SearchServiceProperties> GetSearchServiceFromAzureAsync(Servi
             url = QueryHelpers.AddQueryString(url, "ignoreFilter", "true");
             url = QueryHelpers.AddQueryString(url, "semVerLevel", "2.0.0");
 
-            var duration = Stopwatch.StartNew();
+            // Determine if there is a cached response and if it is considered complete.
+            V2SearchResponse response;
             var complete = false;
-            do
+            var cached = false;
+            lock (_pollCacheLock)
```

Comment:
Consider using a semaphore for async lock

Reply:
Done.